.github/actions/update-extension-provider-images/update-images.py

```diff
@@ -0,0 +1,626 @@
+#!/usr/bin/env python3
+#
+# SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+#
+# SPDX-License-Identifier: Apache-2.0
+"""
+Image Updater Script (Reusable GitHub Action)
+
+Purpose:
+This script is part of a reusable GitHub Action located in the same folder.
+It automates the process of checking for and updating container image
+versions defined in Gardener's extension provider repos at `imagevector/images.yaml`.
+
+How it works:
+1.  The action is triggered by a reusable workflow.
+2.  It reads all image entries from the `images.yaml` file.
+3.  It uses the `oci.client` library to query container registries for new tags.
+4.  It updates the `images.yaml` file in-place based on a set of rules:
+    - For patch updates, it updates the tag of the existing entry.
+    - For new minor/major versions, it adds a new image entry.
+    - It ensures the highest minor version has a `targetVersion` of ">= X.Y".
+    - For images without a `targetVersion`, it keeps only the single latest tag.
+5.  It generates a `release-notes.md` file containing direct links to the
+    release pages for all new and intermediate tags.
+
+This script is not intended to be run manually by developers. It should be
+invoked via the `gardener/cc-utils/.github/workflows/update-extension-provider-images.yaml`
+reusable workflow.
+
+Manual Intervention:
+This script handles version *updates*. Manually editing `images.yaml` is still
+required to remove entire image groups for deprecated Kubernetes versions that
+are no longer supported.
+Image Updater Script.
+"""
+
+import argparse
+import yaml
+import re
+import sys
+import semver
+import time
+from typing import Dict, List, Tuple, Optional, TypedDict
+from collections import defaultdict
+from oci import client as oci_client
+from oci.model import OciImageNotFoundException
+from version import parse_to_semver, is_final
+
+
+# --- Type Definitions ---
+class Update(TypedDict):
+    """A structured representation of a single image update."""
+
+    image_name: str
+    old_tag: Optional[str]
```

Comment:
According to [documentation](https://typing.python.org/en/latest/spec/typeddict.html#required-and-notrequired), you should use Required/NotRequired for that. I think.

Note that we typically use [dataclasses](https://docs.python.org/3/library/dataclasses.html) in cc-utils (there, `Optional` is appropriate to use :-))